contracts/src/WarpProtocolRegistry.sol

```diff
@@ -0,0 +1,181 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+
+/**
+ * @dev Implementation of an abstract `WarpProtocolRegistry` contract.
+ *
+ * This implementation is a contract that can be used as a base contract for protocols that are
+ * build on top of warp. It allows the protocol to be upgraded through a warp out-of-band message.
+ */
+abstract contract WarpProtocolRegistry {
+    // Address that the out-of-band warp message sets as the "source" address.
+    // The address is obviously not owned by any EOA or smart contract account, so it
+    // can not possibly be the source address of any other warp message emitted by the VM.
+    address public constant VALIDATORS_SOURCE_ADDRESS = address(0);
+
+    WarpMessenger public constant WARP_MESSENGER =
+        WarpMessenger(0x0200000000000000000000000000000000000005);
+
+    bytes32 internal immutable _chainID;
+
+    // The latest protocol version. 0 means no protocol version has been added, and isn't a valid version.
+    uint256 internal _latestVersion;
+
+    mapping(uint256 => address) internal _versionToAddress;
+
+    /**
+     * @dev Emitted when a new protocol version is added to the registry.
+     */
+    event AddProtocolVersion(
+        uint256 indexed version,
+        address indexed protocolAddress
+    );
+
+    // Errors
+    error InvalidWarpMessage();
+    error InvalidSourceChainID();
+    error InvalidOriginSenderAddress();
+    error InvalidDestinationChainID();
+    error InvalidDestinationAddress();
+    error InvalidProtocolAddress();
+    error InvalidProtocolVersion();
+    error InvalidRegistryInitialization();
+
+    /**
+     * @dev Initializes the contract by setting a `chainID` and `latestVersion`.
+     */
+    constructor(
+        uint256[] memory initialVersions,
+        address[] memory initialProtocolAddresses
```

Comment:
Could we combine these into an array of structs to eliminate index-mismatch user errors?

Reply:
changed to use `ProtocolRegistryEntry` to avoid index mismatch, good point.